firedrake/variational_solver.py

```diff
@@ -34,43 +39,128 @@ def __init__(self, F, u, bcs=None, J=None,
-        self.J = J or ufl_expr.derivative(F, u)
 
-        if not isinstance(self.J, (ufl.Form, slate.slate.TensorBase)):
-            raise TypeError("Provided Jacobian is a '%s', not a Form or Slate Tensor" % type(self.J).__name__)
-        if len(self.J.arguments()) != 2:
-            raise ValueError("Provided Jacobian is not a bilinear form")
-        if self.Jp is not None and not isinstance(self.Jp, (ufl.Form, slate.slate.TensorBase)):
-            raise TypeError("Provided preconditioner is a '%s', not a Form or Slate Tensor" % type(self.Jp).__name__)
-        if self.Jp is not None and len(self.Jp.arguments()) != 2:
-            raise ValueError("Provided preconditioner is not a bilinear form")
+        # Domain form
+        # linear
+        if isinstance(eq.lhs, ufl.Form) and isinstance(eq.rhs, ufl.Form):
+            self.J = eq.lhs
+            self.Jp = Jp
+            if eq.rhs is 0:
+                self.F = ufl_expr.action(self.J, u)
+            else:
+                if not isinstance(eq.rhs, (ufl.Form, slate.slate.TensorBase)):
+                    raise TypeError("Provided RHS is a '%s', not a Form or Slate Tensor" % type(eq.rhs).__name__)
+                if len(eq.rhs.arguments()) != 1:
+                    raise ValueError("Provided RHS is not a linear form")
+                self.F = ufl_expr.action(self.J, u) - eq.rhs
+        # nonlinear
+        else:
+            if eq.rhs != 0:
+                raise TypeError("r.h.s. of nonlinear form has to be 0")
+            if not isinstance(eq.lhs, (ufl.Form, slate.slate.TensorBase)):
+                raise TypeError("Provided residual is a '%s', not a Form or Slate Tensor" % type(eq.lhs).__name__)
+            if len(eq.lhs.arguments()) != 1:
+                raise ValueError("Provided residual is not a linear form")
+            self.F = eq.lhs
+            self.J = J or ufl_expr.derivative(self.F, u)
+            self.Jp = Jp
+            if not isinstance(self.J, (ufl.Form, slate.slate.TensorBase)):
+                raise TypeError("Provided Jacobian is a '%s', not a Form or Slate Tensor" % type(self.J).__name__)
+            if len(self.J.arguments()) != 2:
+                raise ValueError("Provided Jacobian is not a bilinear form")
+            if self.Jp is not None and not isinstance(self.Jp, (ufl.Form, slate.slate.TensorBase)):
+                raise TypeError("Provided preconditioner is a '%s', not a Form or Slate Tensor" % type(self.Jp).__name__)
+            if self.Jp is not None and len(self.Jp.arguments()) != 2:
+                raise ValueError("Provided preconditioner is not a bilinear form")
+            self.is_linear = False
+
+        # Boundary (FormBC)
+        if self.bcs is not None:
+            for bc in self.bcs:
+                if isinstance(bc, FormBC):
+                    # linear
+                    if isinstance(bc.formeq.lhs, ufl.Form) and isinstance(bc.formeq.rhs, ufl.Form):
+                        bc.J = bc.formeq.lhs
```

Comment:
Careful, this smells like bad code to me, because you're updating the `FormBC` object in place. In particular, I think this means you can't use the same `FormBC` object with two different `VariationalProblems`, since the `bc.J`, etc... will be set by the most recently constructed one.

Reply:
I will move this part to the constructor, too.